### PR TITLE
Fix #33427

### LINKED ISSUE #33427: [Translator] Empty string
**Symfony version(s) affected**: 4.3.4

**Description**  
We have a service that calls the translator with dynamically created key and we sometimes have to use empty strings (or the translator will return the key if the key is not found in the translation file).

But using an empty string throws an exception as per https://bugs.php.net/bug.php?id=78484

**How to reproduce**  
```yaml
# test+intl-icu.yaml
empty: ''
```
```php
$translator->trans('fr_FR', 'empty', 'test');
```

Root cause to reproduce the `IntlException: Constructor failed` exception

```php
new MessageFormatter('fr_FR', '');
```

**Possible Solution**  
I don't know if the PHP core team plans to change this behavior but it can easily worked around in `Symfony\Component\Translation\Translator::trans()`: if `$catalogue->get($id, $domain)` returns an empty string then just return an empty string instead of calling `$this->formatter->formatIntl()` or `$this->formatter->format()`.

I can submit a PR is needed

### PR DESCRIPTION
**Pull request by sylfabre**
| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #33427
| License       | MIT
| Doc PR        | none

Fix #33427 by checking if the message returned by the intl-icu catalog is empty. If yes then the translator returns an empty string instead of running `formatIntl()` which uses the constructor of `MessageFormatter` which throws an exception with empty strings.

### CODE REVIEW

File: src/Symfony/Component/Translation/Translator.php [hunk 1]
```diff
@@ -213,7 +213,13 @@ public function trans($id, array $parameters = [], $domain = null, $locale = nul
         }
 
         if ($this->hasIntlFormatter && $catalogue->defines($id, $domain.MessageCatalogue::INTL_DOMAIN_SUFFIX)) {
-            return $this->formatter->formatIntl($catalogue->get($id, $domain), $locale, $parameters);
+            $message = $catalogue->get($id, $domain);
+            // formatIntl uses MessageFormatter which constructor throws an exception if the message is empty
+            if ('' === $message) {
+                return '';
+            } else {
```
Comment (stof) [inline src/Symfony/Component/Translation/Translator.php]: don't use `else` here, as the `if` returns.

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Translation/Translator.php (hunk 1 above)
Comment (sylfabre) [reply src/Symfony/Component/Translation/Translator.php]: @stof fixed

File: src/Symfony/Component/Translation/Tests/TranslatorTest.php [hunk 2]
```diff
@@ -449,6 +449,7 @@ public function getTransFileTests()
     public function getTransTests()
     {
         return [
+            ['', 'Empty string', '', [], 'fr', ''],
```
Comment (stof) [inline src/Symfony/Component/Translation/Tests/TranslatorTest.php]: isn't this testing it for the non-intl version ? Other messages are for the legacy format here.

Comment (sylfabre) [reply src/Symfony/Component/Translation/Tests/TranslatorTest.php]: @stof this tests that there no exception is thrown with an empty message when you don't use `+intl-icu`. It was already working fine, I've just added a test for this corner case.

File: src/Symfony/Component/Translation/Translator.php [hunk 3]
```diff
@@ -213,7 +213,13 @@ public function trans($id, array $parameters = [], $domain = null, $locale = nul
         }
 
         if ($this->hasIntlFormatter && $catalogue->defines($id, $domain.MessageCatalogue::INTL_DOMAIN_SUFFIX)) {
-            return $this->formatter->formatIntl($catalogue->get($id, $domain), $locale, $parameters);
+            $message = $catalogue->get($id, $domain);
+            // formatIntl uses MessageFormatter which constructor throws an exception if the message is empty
+            if ('' === $message) {
+                return '';
+            }
```
Comment (yceruto) [inline src/Symfony/Component/Translation/Translator.php]: I think this check should be moved to `IntlFormatter::formatIntl()` method, otherwise anyone using it will face the same problem.

**Review by yceruto — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Translation/Tests/TranslatorTest.php (hunk 2 above)
Comment (sylfabre) [reply src/Symfony/Component/Translation/Tests/TranslatorTest.php]: According to @yceruto comment, I'll update the PR so this comment is no longer relevant

**Review by yceruto — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @sylfabre.
